src/argparse.zig

```diff
@@ -0,0 +1,297 @@
+						return .{.failure = .{.message = message}};
+					};
+				}
+			}
+
+			if(split.next() != null) {
+				const message = std.fmt.allocPrint(allocator.allocator, "Too many arguments for command, expected {}", .{count}) catch unreachable;
+				return .{.failure = .{.message = message}};
+			}
+
+			return .{.success = result};
+		}
+		fn parseArgument(comptime Field: type, _arg: ?[]const u8) !Field {
+			switch(@typeInfo(Field)) {
+				inline .optional => |optionalInfo| {
+					if(_arg == null) return null;
+					return try parseArgument(optionalInfo.child, _arg);
+				},
+				else => |fieldInfo| {
+					const arg = _arg orelse return error.MissingArgument;
+					switch(fieldInfo) {
+						inline .@"struct" => {
+							if(!@hasDecl(Field, "parse")) @compileError("Struct must have a parse function");
+							return try @field(Field, "parse")(arg);
+						},
+						inline .@"enum" => return std.meta.stringToEnum(Field, arg) orelse return error.InvalidEnum,
+						inline .float => |floatInfo| return try std.fmt.parseFloat(std.meta.Float(floatInfo.bits), arg),
+						inline .int => |intInfo| return try std.fmt.parseInt(std.meta.Int(intInfo.signedness, intInfo.bits), arg, 0),
+						inline else => |other| @compileError("Unsupported type " ++ @tagName(other)),
+					}
+				},
+			}
+		}
+		fn parseUnion(comptime u: std.builtin.Type.Union, allocator: NeverFailingAllocator, args: []const u8) ParseResult(T) {
+			var failureMessages: ListUnmanaged([]const u8) = .{};
+			defer {
+				for(failureMessages.items[1..]) |item| {
+					allocator.free(item);
+				}
+				failureMessages.deinit(allocator);
+			}
+
+			failureMessages.ensureCapacity(allocator, u.fields.len + 1);
+			failureMessages.appendAssumeCapacity("Provided argument list didn't match any of the valid alternative interpretations of command argument list.");
+
+			inline for(u.fields) |field| {
+				const fieldResult = Parser(field.type, null)._parse(allocator, args);
+				if(fieldResult == .success) {
+					return .{.success = @unionInit(T, field.name, fieldResult.success)};
+				}
+				failureMessages.appendAssumeCapacity(fieldResult.failure.message);
+			}
+
+			const message = join(allocator, '\n', failureMessages.items);
+			return .{.failure = .{.message = message}};
+		}
+	};
+}
+
+pub fn join(allocator: NeverFailingAllocator, char: u8, args: [][]const u8) []const u8 {
+	var totalLength: usize = 0;
+	for(args) |arg| {
+		totalLength += arg.len;
+	}
+	totalLength += args.len - 1;
+
+	const result = allocator.alloc(u8, totalLength);
+	var offset: usize = 0;
+
+	for(args) |arg| {
+		@memcpy(result[offset .. offset + arg.len], arg);
+		offset += arg.len;
+		if(offset < totalLength) {
+			result[offset] = char;
+			offset += 1;
+		}
+	}
+
+	return result;
+}
+
+pub fn ParseResult(comptime SuccessT: type) type {
+	return union(enum) {
+		failure: struct {
+			message: []const u8,
+		},
+		success: SuccessT,
+
+		pub fn deinit(self: @This(), allocator: NeverFailingAllocator) void {
```

Comment:
Hmm, this function does seem quite easy to miss. It seems you missed it in the reference implementation too.

Maybe a better pattern would be to pass in the result message List. This would allow using regular errors, with their syntax sugar, as well as moving this up one layer into the command code, so we could do the error message handling there. Either way, that would be a bigger rewrite, and doesn't need to be done here.

Reply:
Noted, but I don't fully understand the request unfortunately, I guess we will return to this topic. What I would like to preserve is the fact that this parser can be used both for commands and for CLI parameters for Cubyz executable if useful at some point.